### PR TITLE
[docs] Add esbuild for mdx transpiling

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This is just a test to see if we can speed up MDX transpiling with esbuild. For now, it would only apply for our MDX files but they make up for _almost_ everything in our docs 😄 

# How

- Based on https://github.com/tdjsnelling/esbuild-loader-nextjs

# Test Plan

- [x] Validate if the JS/HTML output is similar to normal export
> I deployed Jon's PR #12010 with this change to [this test bucket](http://cedrics-expo.s3-website.eu-central-1.amazonaws.com/)

# Local performance tests

Locally, it looks promising. It speeds up the exporting around ±2 times. From `Done in 433.83s.` to `Done in 232.75s.`. Let's see how fast it performs in CI.

Edit: CI seems to be ±1.53x faster, not a dramatic increase. From 10:27 to 6:50.

Edit2: Both Anav and Cruzan see around the x2 performance increment locally (which is great!)

### CODE REVIEW

File: docs/next.config.js [hunk 1]
```diff
@@ -18,14 +19,23 @@ const latest = join('pages', 'versions', 'latest/');
...
+
 module.exports = {
   trailingSlash: true,
   // Rather than use `@zeit/next-mdx`, we replicate it
   pageExtensions: ['js', 'jsx', 'ts', 'tsx', 'md', 'mdx'],
   webpack: (config, options) => {
     // Add preval support for `constants/*` only and move it to the `.next/preval` cache.
     // It's to prevent over-usage and separate the cache to allow manually invalidation.
-    // See: https://github.com/kentcdodds/babel-plugin-preval/issues/19
+    // See: https://github.com/kentcdodds/babel-plugin-preval/issues/19\
```
Comment (ide) [inline docs/next.config.js:38]:
```suggestion
    // See: https://github.com/kentcdodds/babel-plugin-preval/issues/19
```

**Comment by byCedric** [comment]
I also managed to get emotion's jsx pragma to play nice with esbuild. I couldn't successfully export the whole app because of weird behavior with this fix during statically rendering, but I did get a sense of performance gain.

- Only rendering MDX pages with esbuild + mdx-loader (which still uses babel) completes in `232.75s` locally
- Rendering everything with esbuild + preval (also babel plugin) completes in `197.59s` locally (with an unsuccessful last step)

So far I think the performance we get from switching everything to esbuild isn't worth it. We have to write _a lot of custom code_ for Next itself, while we only gain ±30s. Only handling MDX pages with esbuild seems like the best option for now. 

What do you think @ide?

**Comment by ide** [comment]
Preprocessing only MDX seems better. It's more encapsulated.

**Review by ide — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]
Nice! 👏 We're getting faster every day 😄

**Pull request merged**